plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Transform/Cse.hs

```diff
@@ -0,0 +1,398 @@
+{-# LANGUAGE LambdaCase        #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TupleSections     #-}
+{-# LANGUAGE TypeApplications  #-}
+{-# LANGUAGE TypeFamilies      #-}
+{-# LANGUAGE TypeOperators     #-}
+
+module UntypedPlutusCore.Transform.Cse (cse) where
+
+import PlutusCore (MonadQuote, Name, freshName)
+import UntypedPlutusCore.Core
+import UntypedPlutusCore.Purity (isWorkFree)
+import UntypedPlutusCore.Size
+
+import Control.Lens
+import Control.Monad
+import Control.Monad.Trans.Class
+import Control.Monad.Trans.Reader
+import Control.Monad.Trans.State.Strict
+import Data.Foldable
+import Data.Hashable
+import Data.HashMap.Strict (HashMap)
+import Data.HashMap.Strict qualified as Map
+import Data.List.Extra
+import Data.Traversable
+
+{- Note [CSE]
+
+-------------------------------------------------------------------------------
+1. Simplifications
+-------------------------------------------------------------------------------
+
+This is a simplified (i.e., not fully optimal) implementation of CSE. The two simplifications
+we made are:
+
+- No alpha equivalence check, i.e., `\x -> x` and `\y -> y` are considered different expressions.
```

Comment:
Can we simply compute the hash after renaming the expression? Haven't reviewed the algorithm description yet.

Reply:
Alpha-renaming may perhaps bring a huge benefit. But personally, I am ok to bring it later on a separate PR

Reply:
And it is difficult to do for several reasons: (1) what about non-closed terms; (2) substitution is also harder

Reply:
> I don't understand why though, I find our type checking strategy convoluted for no good reason, but perhaps I'm missing something

Didn't you write the typechecker? :p

Reply:
I basically rewrote it, not just wrote it, and I was opposed to using global uniqueness there from before the beginning:

![Screenshot from 2023-11-16 11-45-15](https://github.com/input-output-hk/plutus/assets/10480926/8dd9c93d-f6a8-42d2-8261-54f309e0198a)

It took me a ridiculous amount of time to weed all the bugs out of the original implementation and to this day I still don't understand why we need global uniqueness there instead of having type values at the type level and reducing them using an abstract machine or whatever like people do.

Reply:
Another thing that I forgot to comment on:

> At least in this implementation, it doesn't CSE work-free expressions anyway.

Such an expression isn't necessarily work-free! You may have `f (\x -> x)` for example.